src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java

```diff
@@ -0,0 +1,213 @@
+package org.jenkins.plugins.lockableresources;
+
+import static java.util.logging.Level.WARNING;
+
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import javax.annotation.Nonnull;
+
+import org.jenkins.plugins.lockableresources.queue.LockableResourcesStruct;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepExecutionImpl;
+import org.jenkinsci.plugins.workflow.steps.BodyExecution;
+import org.jenkinsci.plugins.workflow.steps.BodyExecutionCallback;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+
+import com.google.common.base.Function;
+import com.google.inject.Inject;
+
+import hudson.model.Executor;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import jenkins.model.CauseOfInterruption;
+import jenkins.util.Timer;
+
+public class LockStepExecution extends AbstractStepExecutionImpl {
+
+	@Inject(optional = true)
+	private LockStep step;
+
+	@StepContextParameter
+	private transient Run<?, ?> run;
+
+	@StepContextParameter
+	private transient TaskListener listener;
+
+	private transient volatile ScheduledFuture<?> task;
+	private volatile BodyExecution body;
+	private final String id = UUID.randomUUID().toString();
+	private static final Logger LOGGER = Logger.getLogger(LockStepExecution.class.getName());
+
+	@Override
+	public boolean start() throws Exception {
+		listener.getLogger().println("Trying to acquire lock on [" + step.resource + "]");
+		if (!lockAndProceed()) {
+			listener.getLogger().println(LockableResourcesManager.get().getLockCause(step.resource));
+			listener.getLogger().println("Waiting for lock...");
+			tryLock(0);
+		}
+		return false;
+	}
+
+	@Override
+	public void onResume() {
+		super.onResume();
+		//LockableResourcesManager.get().load();
+		if (body == null) {
+			// Restarted while waiting to lock the resource
+			LOGGER.fine("Resuming lock step on [" + run.getExternalizableId() + "], retrying to acquire lock");
+			tryLock(0);
+		}
+		// the body was already started, nothing to do here
+	}
+
+	private void tryLock(long delay) {
+		task = Timer.get().schedule(new Runnable() {
+			@Override
+			public void run() {
+				task = null;
+				if (!lockAndProceed()) {
+					retry(id, 1000); // try to lock every second
```

Comment:
Does not make sense here. When a lock is released, there should be some sort of notification inside this plugin, and the paused step should be unblocked. There is no need to “poll” for a lock defined inside Jenkins itself.